File: haskell/bob/Bob.hs
module Bob (responseFor) where

import Data.List
import Data.Char

responseFor :: String -> String

isSilence :: String -> Bool
isScreaming :: String -> Bool
isAsking :: String -> Bool

responseFor phrase
  | isSilence phrase   = "Fine. Be that way!"
  | isScreaming phrase = "Whoa, chill out!"
  | isAsking phrase    = "Sure."
  | otherwise          = "Whatever."

isSilence = null . dropWhile isSpace

isScreaming phrase =  letters /= "" && all isUpper letters  
                      where letters = filter isLetter phrase

isAsking = isSuffixOf "?"

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
        question?(input) -> "Sure."
        yell?(input) -> "Whoa, chill out!"
        nothing?(input) -> "Fine. Be that way!"
        true -> "Whatever."
    end
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end


  defp yell?(input) do
    String.upcase(input) == input && String.match?(input, ~r/\p{L}/)
  end

  defp nothing?(input) do
    String.trim(input) == ""
  end

end
